src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,432 @@
+                            /* notModified= */ true,
+                            SystemClock.elapsedRealtime() - requestStartMs,
+                            combinedHeaders));
+            return;
+        }
+
+        byte[] responseContents = httpResponse.getContentBytes();
+
+        if (responseContents == null) {
+            final InputStream inputStream = httpResponse.getContent();
+            if (inputStream == null) {
+                // Add 0 byte response as a way of honestly representing a
+                // no-content request.
+                responseContents = new byte[0];
+            } else {
+                Runnable run =
+                        new Runnable() {
+                            @Override
+                            public void run() {
+                                PoolingByteArrayOutputStream bytes =
+                                        new PoolingByteArrayOutputStream(
+                                                mPool, httpResponse.getContentLength());
+                                byte[] buffer = new byte[1024];
+                                int count = 0;
+                                while (true) {
+                                    try {
+                                        if ((count = inputStream.read(buffer)) == -1) break;
+                                    } catch (IOException e) {
+                                        onRequestFailed(
+                                                request,
+                                                callback,
+                                                e,
+                                                requestStartMs,
+                                                httpResponse,
+                                                bytes.toByteArray());
+                                    }
+                                    bytes.write(buffer, 0, count);
+                                }
+                                byte[] finalResponseContents = bytes.toByteArray();
+                                runAfterBytesReceived(
+                                        requestStartMs,
+                                        statusCode,
+                                        httpResponse,
+                                        request,
+                                        callback,
+                                        responseHeaders,
+                                        finalResponseContents);
+                            }
+                        };
+                mBlockingExecutor.execute(run);
+                return;
+            }
+        }
+        runAfterBytesReceived(
+                requestStartMs,
+                statusCode,
+                httpResponse,
+                request,
+                callback,
+                responseHeaders,
+                responseContents);
+    }
+
+    /* Method to be called after a failed network request */
+    private void onRequestFailed(
```

Comment:
any more code sharing we can do here w/ BasicNetwork to go from Exception type to how it gets handled?